### ISSUE TITLE
How to ban some parameter subspace?

### ISSUE THREAD

**Issue by WeirdKeksButtonK** [optuna/optuna#1134]
Let's say I want to tune [SVM](https://scikit-learn.org/stable/modules/generated/sklearn.svm.LinearSVC.html).

This is pseudo-code for the parameter space:
```python
C: loguniform(0.3, 10.0)
penalty: categorical(['l1', 'l2'])
loss: categorical(['hinge', 'squared_hinge'])
```
The problem is that the combination `penalty='l1' AND loss='hinge'` is not supported by scikit-learn. Can I tell Optuna not to sample this combination?

I tried raising an exception (used the `catch` argument), but Optuna constantly samples this combination anyway. Returning some bad score is not valid from the optimization perspective. One more option is to make one categorical space consisting of valid pairs, but this is both not completely valid and inconvenient. I can also firstly sample penalty and conditionally change the space for loss, but I am not sure if it is valid  from the optimization perspective (as far as I can see from FAQ, in the case of TPE, which I use, it is more or less ok to do so).

**Comment by crcrpar** [comment]
Hi @WeirdKeksButtonK 

How about defining `loss` space conditionally as described in [Branches and Loops](https://optuna.readthedocs.io/en/stable/tutorial/configurations.html#branches-and-loops)? For example,

```python
def objective(trial):
    ...
    C = trial.suggest_loguniform("C", 0.3, 10.0)
    penalty = trial.suggest_categorical("penalty", ["l1", "l2"])
    if penalty == "l1":
        # I guess `loss = "squared_hinge"` is also fine.
        loss = trial.suggest_categorical("loss_for_l1_penalty", ["squared_hinge"])
    else:
        loss = trial.suggest_categorical("loss_for_l2_penalty", ["hinge", "squared_hinge"])
    ...
```

**Comment by ytsmiling** [comment]
Hi,
You can use python to describe the constraint. Does the following code work for your case?

```python
def objective(trial):
    c = trial.suggest_loguniform('C', 0.3, 10.0)
    penalty = trial.suggest_categorical(['l1', 'l2'])
    if penalty == 'l1':
        loss = 'hinge'
    else:
        loss = trial.suggest_categorical('loss-l2', ['hinge', 'squared_hinge'])
    return your_target_func(c=c, penalty=penalty, loss=loss)
```

**Comment by WeirdKeksButtonK** [comment]
Yes, this is my current solution. However, if `penalty=='l1'` then `loss='hinge'` is "sampled" without the sampler knowing it, hence, without updating internal statistics of the sampler related to `loss`. 

In general form:
```python
if condition:
    a = trial.suggest('a_if_true', ...)
else:
    a = trial.suggest('a_if_false', ...)
```
In this case, statistics of `a_if_true` and `a_if_false` are not shared. In theory, this is fully valid, but in practice one could (depends on a task) benefit if the statistics were shared. In other words, it is all about dynamic categorical space. As I learnt from experience, currently, it is not supported for TPE.

**Comment by WeirdKeksButtonK** [comment]
Probably, it was too early to close the issue, I don't know if there are plans to support dynamic categorical spaces for TPE. For me, it is not a blocker, anyway.
